text/0000-driver-model.md

```diff
@@ -0,0 +1,328 @@
+- Feature Name: driver_model
+- Start Date: 2022-07-15
+- RFC PR: [twizzler-rfcs/rfcs#0000](https://github.com/twizzler-operating-system/rfcs/pull/0000)
+- Twizzler Issue: [twizzler-operating-system/twizzler#0000](https://github.com/twizzler-operating-system/twizzler/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC extends RFCxxxx with a set of higher-level abstractions that drivers can use to program
+devices. It describes centralized abstractions surrounding event streams, request-response queues, asynchrony,
+and interrupts that will be provided by the twizzler-driver crate.
+
+# Motivation
+[motivation]: #motivation
+
+Writing device drivers is hard, in no small part due to complex models inherent to the asynchronous
+nature of interacting with the devices. Developers can make use of infrastructure surrounding device
+drivers in most operating systems to lessen this complexity. Devices can be (often, largely) modeled
+as separate computing devices that a) produce events for host software to consume, and b) receive
+and respond to commands sent by host software. Thus we want to provide a mechanism that allows these
+abstractions to be implemented by a majority of device drivers easily while providing a unified
+framework that higher-level aspects of a driver can use.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The twizzler-driver crate provides a high-level wrapper around a collection of abstractions designed
+to manage a single device (for definition of "device" and the `Device` type, see RFCxxxx), called a
+`DeviceController`. When taking control of a device, driver software can create a `DeviceController`
+(heretoafter referred to as "the controller") from a `Device`. After this point, the controller
+manages the device and exposes several abstractions above the device:
+
+ - A `DeviceEventStream`, which provides a way to get a stream of events, including mailbox events
+   and device interrupts (see RFCxxxx).
+ - A `DmaAllocator` (details for which will appear in a future RFC).
+ - A `RequestEngineManager`, which provides a way to submit requests to the device and
+   asynchronously await responses.
+
+## Device Event Stream
+
+The device event stream provides an async interface for handling (higher-half) interrupts and
+mailbox messages. This means that some async executor is required; if this is too much overhead for
+the application, direct access to the underlying Device allows for lower-level interrupt handling
+without the async overhead.
+
+This allows driver software to handle incoming events from the device. For example, a NIC that
```

Comment:
Annoying person that I am.... Dont' start sentences with "This" always start with whatever This actually is.  McKusick beat that into me on our first go round of D&I.